src/main/java/ai/promoted/delivery/client/DeliveryRequestState.java

```diff
@@ -0,0 +1,69 @@
+package ai.promoted.delivery.client;
+
+import java.util.List;
+import ai.promoted.delivery.model.Insertion;
+import ai.promoted.delivery.model.Request;
+import ai.promoted.delivery.model.Response;
+
+/**
+ * Holds state of a request during processing.
+ */
+class DeliveryRequestState {
+
+  /** The delivery request we're processing */
+  private final DeliveryRequest deliveryRequest;
+
+  /** Any insertions on the request that were not passed to Delivery API */
+  private List<Insertion> nonApiInsertions;
+  
+  /**
+   * Creates a state object for the given request.
+   * 
+   * @param deliveryRequest the delivery request we are processing.
+   */
+  public DeliveryRequestState(DeliveryRequest deliveryRequest) {
+    this.deliveryRequest = deliveryRequest;
+  }
+  
+  /**
+   * Gets the response to return to the client with any necessary post-processing.
+   * @param resp the response from Delivery API.
+   * @return the post-processed response
+   */
+  public Response getResponseToReturn(Response resp) {
+    
+    // Add any insertions back past the maximum, and assign them an insertion id.
```

Comment:
Before we do this approach

1. I'd have to verify that our Flink system will join these correctly.  I don't think it does currently.  I updated the KI with the following.

> @dhill - Need to update Flink join to handle a mix of API and SDK DeliveryLog insertions.

2. if we generate insertionId on the client-side, we'd have to log a DeliveryLog for it (or create a new log record).


For now, I'd choose to either hard truncate the results or not return an insertionId (which would fall back to inferred references).  This might confuse customers though.  Having an insertionId that we don't know about might cause weird DQ issues (since we'll assume that we're missing DeliverLogs).